index.js

```diff
@@ -393,86 +457,144 @@ Compiler.prototype = {
-          }
-          this.buf.push('attributes: ' + this.runtime('merge') + '([' + attrsBlocks.join(',') + '])');
-        } else if (attrs.length) {
-          var val = this.attrs(attrs);
-          this.buf.push('attributes: ' + val);
-        }
-
-        if (args) {
-          this.buf.push('}, ' + args + ');');
+          this.buf.push('},');
         } else {
-          this.buf.push('});');
+          this.buf.push('}');
         }
+      }
 
-      } else {
-        this.buf.push(name + '(' + args + ');');
+      if (attrsBlocks.length === 1) {
+        this.buf.push('attributes: ' + attrsBlocks[0]);
+      } else if (attrsBlocks.length) {
+        this.buf.push('attributes: ' + this.runtime('merge') + '([' + attrsBlocks.join(',') + '])');
       }
-      if (pp) this.buf.push("jade_indent.pop();")
+      this.buf.push('}');
+      args.unshift(stack().join(''));
     } else {
-      var mixin_start = this.buf.length;
-      args = args ? args.split(',') : [];
-      var rest;
-      if (args.length && /^\.\.\./.test(args[args.length - 1].trim())) {
-        rest = args.pop().trim().replace(/^\.\.\./, '');
-      }
-      this.buf.push(name + ' = function(' + args.join(',') + '){');
-      this.buf.push('var block = (this && this.block), attributes = (this && this.attributes) || {};');
-      if (rest) {
-        this.buf.push('var ' + rest + ' = [];');
-        this.buf.push('for (jade_interp = ' + args.length + '; jade_interp < arguments.length; jade_interp++) {');
-        this.buf.push('  ' + rest + '.push(arguments[jade_interp]);');
-        this.buf.push('}');
-      }
-      this.parentIndents++;
-      this.visit(block);
-      this.parentIndents--;
-      this.buf.push('};');
-      var mixin_end = this.buf.length;
-      this.mixins[key].instances.push({start: mixin_start, end: mixin_end});
+      args.unshift('null');
+    }
+    if (dynamic) {
+      this.bufferExpression('jade_mixins[' + mixin.name.substr(2,mixin.name.length-3) + '](' + args.join(',') + ')');
+    } else {
+      this.bufferExpression(this.delayedContent(function () {
+        var mixinsWithName = Object.keys(this.mixins).filter(function (key) {
+          return this.mixins[key].names.indexOf(mixin.name) !== -1;
+        }.bind(this));
+        if (mixinsWithName.length === 1) {
+          return mixinsWithName[0];
+        } else if (mixinsWithName.length > 1) {
+          return 'jade_mixins["' + mixin.name + '"]';
+        } else {
+          return 'function () { throw new Error("Mixin \\"' + mixin.name + '\\" is not defined"); }';
+        }
+      }) + '(' + args.join(',') + ')');
+    }
+    if (pp) this.buf.push("jade_indent.pop();")
+  },
+  visitMixinDefinition: function (mixin) {
+    var originalCalledMixins = this.calledMixins;
+    this.calledMixins = [];
+    var args = mixin.args ? mixin.args.split(',') : [];
+    args.unshift('jade_mixin_data');
+    var rest;
+    if (/^\.\.\../.test(args[args.length - 1].trim())) {
```

Comment:
There should be at least some character for the variable name, in addition to the `...` to mark it as a rest param.  We should probably be more restrictive and actually error early if any of these are not valid identifiers.  I don't want to get hung up on that detail at this point though.  The higher level optimisation is much more important.